eclipse/src/saros/editor/internal/EditorAPI.java

```diff
@@ -268,27 +272,275 @@ public static IResource getEditorResource(IEditorPart editorPart) {
+
+    return new TextSelection(startPosition, endPosition);
+  }
+
+  /**
+   * Calculates the text based offsets in the given editor part for the given text selection.
+   *
+   * <p>The given editor part must not be <code>null</code> and the given text selection must not be
+   * <code>null</code> and must not be empty.
+   *
+   * @param editorPart the local editor part
+   * @param selection the text selection
+   * @return the absolute offsets in the given editor part for the given text selection or {@link
+   *     org.eclipse.jface.text.TextSelection#emptySelection()} if the given editor is not a text
+   *     editor, the editor does not have a valid document provider, a valid IDocument could not be
+   *     obtained form the document provider, or the line offsets could not be calculated
+   * @throws NullPointerException if the given editor part or selection is <code>null</code>
+   * @throws IllegalArgumentException if the given text selection is empty
+   */
+  public static ITextSelection calculateOffsets(IEditorPart editorPart, TextSelection selection) {
+
+    Objects.requireNonNull(editorPart, "The given editor part must not be null");
+    Objects.requireNonNull(selection, "The given selection must not be null");
+
+    if (selection.isEmpty()) {
+      throw new IllegalArgumentException("The given text selection must not be empty");
+    }
+
+    if (!(editorPart instanceof ITextEditor)) {
+      return org.eclipse.jface.text.TextSelection.emptySelection();
+    }
+
+    ITextEditor textEditor = (ITextEditor) editorPart;
+
+    IDocumentProvider documentProvider = textEditor.getDocumentProvider();
+
+    if (documentProvider == null) {
+      return org.eclipse.jface.text.TextSelection.emptySelection();
+    }
+
+    IDocument document = textEditor.getDocumentProvider().getDocument(editorPart.getEditorInput());
+
+    if (document == null) {
+      return org.eclipse.jface.text.TextSelection.emptySelection();
     }
 
-    ITextSelection jfaceSelection = (ITextSelection) selectionProvider.getSelection();
+    TextPosition startPosition = selection.getStartPosition();
+    TextPosition endPosition = selection.getEndPosition();
+
+    int startLineOffset;
+    int endLineOffset;
+
+    try {
+      startLineOffset = document.getLineOffset(startPosition.getLineNumber());
+
+      if (endPosition.getLineNumber() == startPosition.getLineNumber()) {
+        endLineOffset = startLineOffset;
+
+      } else {
+        endLineOffset = document.getLineOffset(endPosition.getLineNumber());
+      }
+
+    } catch (BadLocationException e) {
+      LOG.warn(
+          "Could not calculate the line offset for the given selection "
+              + selection
+              + " in the document "
+              + document,
+          e);
+
+      return org.eclipse.jface.text.TextSelection.emptySelection();
+    }
+
+    int startOffset = startLineOffset + startPosition.getInLineOffset();
```

Comment:
You should add a sanity check because there is now way compare TextLocations.

Eg. You can have a TextLocation(0, 20) and TextLocation(1, 5). The first TextLocation is invalid because the first line was altered in some way only only contains 4 characters now. In the end you will end up with a negative length. This can also happen if you just transmit faulty values.

Reply:
Yes, I agree that a sanity check would be nice when calculating the selection values. I will think about how best to do this.

But I also don't think that this is an issue new to then new handling. As selections are not processed as part of the jupiter operation stuff, we could always run into situations where a local selection can no longer be applied correctly as the document has changed.

Reply:
My guess is that the annotation position and or offset does not really have to match the current document boundaries the annotation is put in. As long as no exception is triggered I am fine with some log statements (warning or debug). You already mentioned that normally selection must also use the Jupiter algorithm but we do not do that.

Reply:
An additional sanity check here should not be necessary. If the given offsets/text positions can no longer be converted, the logic throws a `BadLocationException`. This exception is already caught internally, logged and an empty selection is returned instead.

And, in the unlikely situation that the line offset is valid but the in-line offset pushes the calculated offset out of bounds, the annotation will be dropped silently when trying to apply it to the local model. See [Eclipse documentation](https://help.eclipse.org/2018-09/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2Fsource%2Fclass-use%2FIAnnotationModelExtension.html) (used in our code [here](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/editor/internal/AnnotationModelHelper.java#L65)).

For IntelliJ, such cases are [also handled by the annotation](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationManager.java#L713) logic.